File: graph/PrimMinimumSpanningTree/java/PrimMST.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

public final class PrimMST {
    private PrimMST() {}

    public static final class Edge {
        public final int source;
        public final int target;
        public final long weight;

        public Edge(int source, int target, long weight) {
            this.source = source;
            this.target = target;
            this.weight = weight;
        }
    }

    private static final class Candidate {
        final long weight;
        final int source;
        final int target;
        final int nextNode;

        Candidate(long weight, int source, int target, int nextNode) {
            this.weight = weight;
            this.source = source;
            this.target = target;
            this.nextNode = nextNode;
        }
    }

    public enum Status {
        OK,
        DISCONNECTED,
        OVERFLOW
    }

    public static final class Result {
        public final Status status;
        public final long totalWeight;
        public final List<Edge> chosenEdges;

        public Result(Status status, long totalWeight, List<Edge> chosenEdges) {
            this.status = status;
            this.totalWeight = totalWeight;
            this.chosenEdges = chosenEdges;
        }
    }

    private static Long checkedAdd(long left, long right) {
        if (right > 0 && left > Long.MAX_VALUE - right) {
            return null;
        }
        if (right < 0 && left < Long.MIN_VALUE - right) {
            return null;
        }
        return left + right;
    }

    private static Edge normalize(Edge edge) {
        if (edge.source <= edge.target) {
            return edge;
        }
        return new Edge(edge.target, edge.source, edge.weight);
    }

    public static Result primMST(int nodeCount, List<Edge> edges) {
        if (nodeCount <= 1) {
            return new Result(Status.OK, 0L, new ArrayList<>());
        }

        List<List<Edge>> adjacency = new ArrayList<>(nodeCount);
        for (int node = 0; node < nodeCount; ++node) {
            adjacency.add(new ArrayList<>());
        }

        for (Edge edge : edges) {
            Edge normalized = normalize(edge);
            adjacency.get(normalized.source).add(new Edge(normalized.source, normalized.target, normalized.weight));
            adjacency.get(normalized.target).add(new Edge(normalized.target, normalized.source, normalized.weight));
        }

        boolean[] visited = new boolean[nodeCount];
        PriorityQueue<Candidate> heap =
                new PriorityQueue<>(
                        Comparator.comparingLong((Candidate candidate) -> candidate.weight)
                                .thenComparingInt(candidate -> candidate.source)
                                .thenComparingInt(candidate -> candidate.target)
                                .thenComparingInt(candidate -> candidate.nextNode));
        List<Edge> chosenEdges = new ArrayList<>(nodeCount - 1);
        long totalWeight = 0L;

        java.util.function.IntConsumer pushCandidates =
                node -> {
                    for (Edge edge : adjacency.get(node)) {
                        if (visited[edge.target]) {
                            continue;
                        }
                        Edge normalized = normalize(new Edge(edge.source, edge.target, edge.weight));
                        heap.add(new Candidate(normalized.weight, normalized.source, normalized.target, edge.target));
                    }
                };

        visited[0] = true;
        int visitedCount = 1;
        pushCandidates.accept(0);

        while (!heap.isEmpty() && visitedCount < nodeCount) {
            Candidate candidate = heap.poll();
            if (visited[candidate.nextNode]) {
                continue;
            }

            Long updated = checkedAdd(totalWeight, candidate.weight);
            if (updated == null) {
                return new Result(Status.OVERFLOW, totalWeight, chosenEdges);
            }

            totalWeight = updated;
            chosenEdges.add(new Edge(candidate.source, candidate.target, candidate.weight));
            visited[candidate.nextNode] = true;
            visitedCount += 1;
            pushCandidates.accept(candidate.nextNode);
        }

        if (visitedCount != nodeCount) {
            return new Result(Status.DISCONNECTED, totalWeight, chosenEdges);
        }

        return new Result(Status.OK, totalWeight, chosenEdges);
    }
}

File: graph/PrimMinimumSpanningTree/python/PrimMST.py
from dataclasses import dataclass
from enum import Enum
import heapq


I64_MIN = -(1 << 63)
I64_MAX = (1 << 63) - 1


@dataclass(frozen=True)
class Edge:
    source: int
    target: int
    weight: int


class PrimStatus(Enum):
    OK = "ok"
    DISCONNECTED = "disconnected"
    OVERFLOW = "overflow"


@dataclass
class PrimResult:
    status: PrimStatus
    total_weight: int
    chosen_edges: list[Edge]


def _checked_add(left: int, right: int) -> int | None:
    total = left + right
    if total < I64_MIN or total > I64_MAX:
        return None
    return total


def _normalize(edge: Edge) -> Edge:
    if edge.source <= edge.target:
        return edge
    return Edge(edge.target, edge.source, edge.weight)


def prim_mst(node_count: int, edges: list[Edge]) -> PrimResult:
    if node_count <= 1:
        return PrimResult(PrimStatus.OK, 0, [])

    adjacency: list[list[Edge]] = [[] for _ in range(node_count)]
    for edge in edges:
        normalized = _normalize(edge)
        adjacency[normalized.source].append(Edge(normalized.source, normalized.target, normalized.weight))
        adjacency[normalized.target].append(Edge(normalized.target, normalized.source, normalized.weight))

    visited = [False] * node_count
    heap: list[tuple[int, int, int, int]] = []
    chosen_edges: list[Edge] = []
    total_weight = 0

    def push_candidates(node: int) -> None:
        for edge in adjacency[node]:
            if visited[edge.target]:
                continue
            normalized = _normalize(Edge(edge.source, edge.target, edge.weight))
            heapq.heappush(heap, (normalized.weight, normalized.source, normalized.target, edge.target))

    visited[0] = True
    visited_count = 1
    push_candidates(0)

    while heap and visited_count < node_count:
        weight, source, target, next_node = heapq.heappop(heap)
        if visited[next_node]:
            continue

        updated = _checked_add(total_weight, weight)
        if updated is None:
            return PrimResult(PrimStatus.OVERFLOW, total_weight, chosen_edges)

        total_weight = updated
        chosen_edges.append(Edge(source, target, weight))
        visited[next_node] = True
        visited_count += 1
        push_candidates(next_node)

    if visited_count != node_count:
        return PrimResult(PrimStatus.DISCONNECTED, total_weight, chosen_edges)

    return PrimResult(PrimStatus.OK, total_weight, chosen_edges)
